### DISCUSSION TITLE
nameof cant handle generic with constraint

### DISCUSSION THREAD

**Discussion by puschie286** [dotnet/csharplang#5494] [General]
### Description

nameof can be used for generic types - but it fails to access interfaces on generic types ( it ignores constraints on generic types ).
This works fine with the actual implementaion

```
public class Example<Template> where Template : ITestValue
{
	public string GetTemplateValueName()
	{
		//return nameof( Template );
		
		return nameof( Template.Test ); // <-- compiler error: 'Template' is a type parameter, which is not valid in the given context
	}
}

public interface ITestValue
{
	public int Test { get; set; }
}
```
dotnet fiddle: https://dotnetfiddle.net/2qIrI8

### Configuration

Checked on .NET 6 and Roslyn 4.0

### Other information

You can foward this to the implementation as a parameter
 `nameof( Implementation.Test )`

**Comment by CyrusNajmabadi** [comment]
This doesn't appear to be wrong.  `Template` is a type parameter, and it doesn't have such a member.   You'd have to propose a language change to allow this.  Note: you can just write `nameof(ITestValue.Test)` instead here.

**Reply by jnm2** [reply]
In fact, there's an argument to be made that `nameof(ITestValue.Test)` is preferred over `nameof(T.Test)` the same way that `nameof(BaseClass.Foo)` is preferred over `nameof(DerivedClass.Foo)`.
